ipq806x: fix CPU_TYPE

Blackyfff opened the pull request:
The Qualcom IPQ806X-Series, aka Snapdragon, aka Krait is not identical with Cortex-A15
see: https://en.wikipedia.org/wiki/Comparison_of_ARMv7-A_cores
this leads to wrong code optimization und undefined behavior.

This may eventually also fix:
https://forum.openwrt.org/t/unexplained-hangs-freezes-and-reboots-with-archer-c2600/61999

Run-Tested: Archer C2600 (gcc 9)

neheb commented:
Why not CPU_TYPE:=cortex-a7 ?

neheb commented:
Interesting...

Anyone know the reason for https://github.com/openwrt/openwrt/blob/master/target/linux/ipq806x/patches-5.4/0065-arm-override-compiler-flags.patch ?

pkgadd commented:
Does that change _really_ (reproducibly) fix the archer c2600 or just eventually? ipq806x has been using cortex a15 since summer 2016 (starting out with cortex a7, followed by a short intermezzo as cortex a9), half a decade ago...

You also need to fix the Signed-of-by.

rsalvaterra commented:
> this leads to wrong code optimization und undefined behavior

Wrong code optimisation? Maybe (GCC doesn't have a Krait pipeline description, as far as I know, Cortex-A15 being the closest one). Undefined behaviour? Impossible, since the memory consistency model is the same.

neheb commented:
@rsalvaterra based on the chart, A15 is more capable than Krait. That implies that some binaries can fail.

Blackyfff commented:
I think all Cortex-A(X) (didn't checked, but A15 for sure) have a slightly enhanced instructions set compared to ARMv7A standard.
I don't overlook what kind of compiler optimizations (now and in future) the compiler applies different for what kind of code. (even for packages)

IMHO it is not good practise to be in good faith reliance because Cortex-A15 is very close.
It's only granted that Krait is fully compatible with ARMv7A.

I don't really know, if the freeze and/or hang problem is solved by this.
From what I can tell, I don't have any freezes or hangs running the device with this commit for 16 days.
But I'll try to let my device run for longer.

I've also compared the openssl and wireguard+routing performance and didn't got any noticable differences.
The kernel recognized processor features are identical, too (and therefore used I believe):

Features        : half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32

Blackyfff committed:
ipq806x: fix CPU_TYPE

The Qualcom IPQ806X-Series, aka Snapdragon, aka Krait is not identical with Cortex-A15
see: https://en.wikipedia.org/wiki/Comparison_of_ARMv7-A_cores
this leads to wrong code optimization und undefined behavior.

This may eventually also fix:
https://forum.openwrt.org/t/unexplained-hangs-freezes-and-reboots-with-archer-c2600/61999

Run-Tested: Archer C2600 (gcc 9)
Signed-off-by: Hannes Märten <freifunk@freifunk-herpf.de>

rsalvaterra commented:
> @rsalvaterra based on the chart, A15 is more capable than Krait. That implies that some binaries can fail.

The chart is incomplete. They only differ in LPAE, whick Krait doesn't support. They're otherwise the same.

rsalvaterra commented:
> Why not CPU_TYPE:=cortex-a7 ?

Because A7 only has 16 floating point registers, while the Krait has 32.

Blackyfff commented:
After digging a little bit deeper it seems like I've misunderstud the correlation, sorry for that.
generic-armv7-a and cortex-a15 flags are mapping to the same architecture 'archv7-a' of course.
I've been stumbled above that and mixed things up.
So if there are no hidden bugs this change couldn't solve the problem, it only has impact to the performance.
Sorry for my confusion.
